src/cpp/Cdr.cpp

```diff
@@ -2754,10 +2879,919 @@ Cdr& Cdr::deserializeWStringSequence(
+{
+    assert(MEMBER_ID_INVALID != current_state.next_member_id_);
+    assert(EncodingAlgorithmFlag::PLAIN_CDR2 == current_encoding_ ||
+            EncodingAlgorithmFlag::DELIMIT_CDR2 == current_encoding_ ||
+            EncodingAlgorithmFlag::PL_CDR2 == current_encoding_);
+
+    if (0 < current_state.member_size_ && EncodingAlgorithmFlag::PL_CDR2 == current_encoding_)
+    {
+        auto last_offset = offset_;
+        set_state(current_state);
+        makeAlign(alignment(sizeof(uint32_t)));
+        if (NO_SERIALIZED_MEMBER_SIZE == serialized_member_size_)
+        {
+            const size_t member_serialized_size = last_offset - offset_ -
+                    (current_state.header_serialized_ == XCdrHeaderSelection::SHORT_HEADER ? 4 : 8);
+            if (8 < member_serialized_size)
+            {
+                switch (current_state.header_serialized_)
+                {
+                    case XCdrHeaderSelection::SHORT_HEADER:
+                        if (AUTO_WITH_SHORT_HEADER_BY_DEFAULT == current_state.header_selection_)
+                        {
+                            xcdr2_change_to_long_member_header(current_state.next_member_id_, member_serialized_size);
+                        }
+                        else
+                        {
+                            throw BadParamException("Cannot encode XCDRv2 LongMemberHeader");
+                        }
+                        break;
+                    case XCdrHeaderSelection::LONG_HEADER:
+                        xcdr2_end_long_member_header(current_state.next_member_id_, member_serialized_size);
+                        break;
+                    default:
+                        assert(false); // header_serialized_ must have only SHORT_HEADER or LONG_HEADER
+                }
+            }
+            else
+            {
+                switch (current_state.header_serialized_)
+                {
+                    case XCdrHeaderSelection::SHORT_HEADER:
+                        xcdr2_end_short_member_header(current_state.next_member_id_, member_serialized_size);
+                        break;
+                    case XCdrHeaderSelection::LONG_HEADER:
+                        if (AUTO_WITH_LONG_HEADER_BY_DEFAULT == current_state.header_selection_)
+                        {
+                            xcdr2_change_to_short_member_header(current_state.next_member_id_, member_serialized_size);
+                        }
+                        else
+                        {
+                            xcdr2_end_long_member_header(current_state.next_member_id_, member_serialized_size);
+                        }
+                        break;
+                    default:
+                        assert(false); // header_serialized_ must have only SHORT_HEADER or LONG_HEADER
+                }
+            }
+
+            jump(member_serialized_size);
+        }
+        else
+        {
+            // Use inner type DHEADER as NEXTINT
```

Comment:
This should be refactored. There are some types that could not use the DHEADER as NEXTINT: short or wchar for instace.